v1.5 - NetworkStreamTransport implementation

JeffCyr opened the pull request:
This pull request includes a new Remoting Transport based on a NetworkStream.

This Transport aims to get the lowest overhead possible on top of .Net Sockets.

A sub-goal is to be able to support easily any Transport based on System.IO.Stream. The PR include two other transports (SslStreamTransport and NamedPipeTransport), but it is still a work in progress.

Initial results are promising, the transport is able to achieve 120K msg/sec on a quad core workstation, compared to ~6K msg/sec with Helios.

The happy path is working, but the code remain largely untested. Other PRs will includes spec until there is 100% code & feature coverage.

(This PR does not include the ActorPath cache to achieve 120K msg/sec, without the patch it is doing ~70K msg/sec)

Aaronontheweb commented on src/core/Akka.Remote/Configuration/Remote.conf:
```diff
@@ -488,6 +488,107 @@ akka {
+      # than remoting protocol expects messages at.
+      # Use "0.0.0.0" to bind to all interfaces.
+      # akka.remote.networkstream.hostname if empty
+      bind-hostname = ""
```
nitpick: we use "public-hostname" to achieve this same thing on the existing Helios transport. I don't have a strong opinion on whether or not those should be consistent.

Aaronontheweb commented on src/core/Akka.Remote/Transport/Streaming/AsyncQueue.cs:
```diff
@@ -0,0 +1,209 @@
+            lock (_queue)
+            {
+                if (_pendingDequeue != null)
+                    throw new InvalidOperationException("Dequeue operation is already in progress. This class is single reader.");
```
Nitpick: I would explain the intention of the design behind this queue in an XML-DOC comment, for the benefit of future contributors who might need to read and modify this code.

Aaronontheweb commented on src/core/Akka.Remote/Transport/Streaming/NamedPipeTransport.cs:
```diff
@@ -0,0 +1,124 @@
+namespace Akka.Remote.Transport.Streaming
+{
+    //TODO Work in progress
+    public class NamedPipeTranspotSettings : StreamTransportSettings
```
Should be `NamedPipeTransportSettings`

Aaronontheweb commented on src/core/Akka.Remote/Transport/Streaming/NetworkStreamTransport.cs:
```diff
@@ -0,0 +1,289 @@
+            else
+            {
+                // If hostname is not localhost or an ip, bind to IPAny
+                BindIp = Socket.OSSupportsIPv6 ? IPAddress.IPv6Any : IPAddress.Any;
```
need to try DNS hostname resolution before binding to `0.0.0.0`

Aaronontheweb replied:
Also: if the user didn't specify a valid hostname OR ipaddress, throw an exception instead. Let them know their configuration is screwed up.

Aaronontheweb replied:
Actually - nevermind. 0.0.0.0 is a fine default - consistent with what JVM Akka and Helios currently do.

Aaronontheweb commented on src/core/Akka.Remote/Transport/Streaming/NetworkStreamTransport.cs:
```diff
@@ -0,0 +1,289 @@
+                bindHostname = Hostname;
+
+            IPAddress ip;
+            if (Hostname.Equals("localhost", StringComparison.OrdinalIgnoreCase))
```
I'd change this to (pseudo-code):

```
if(!IPAddress.TryParse(Hostname, out ip)){
   //dns
   ip = Dns.GetHostEntry(HostName).AddressList[0]; // might want to cache somewhere in settings
} else{
   /* set the BIND IP  */
}
```

Aaronontheweb commented on src/core/Akka.Remote/Transport/Streaming/NetworkStreamTransport.cs:
```diff
@@ -0,0 +1,289 @@
+
+            Port = config.GetInt("port");
+
+            Hostname = config.GetString("hostname");
```
This will override the previous `Hostname` set above - delete.

Aaronontheweb commented on src/core/Akka.Remote/Transport/Streaming/NetworkStreamTransport.cs:
```diff
@@ -0,0 +1,289 @@
+            Port = config.GetInt("port");
+
+            Hostname = config.GetString("hostname");
+            Port = config.GetInt("port");
```
Redundant `Port` set here also.

Aaronontheweb commented on src/core/Akka.Remote/Transport/Streaming/NetworkStreamTransport.cs:
```diff
@@ -0,0 +1,289 @@
+
+    public class NetworkStreamTransport : StreamTransport
+    {
+        public const string ProtocolName = "tcp";
```
Protocol name needs to be configurable via HOCON.

Aaronontheweb replied:
Whoops, nevermind I stand corrected - Helios / Netty only do this because they support UDP and TCP. This is fine otherwise.

Aaronontheweb commented on src/core/Akka.Remote/Transport/Streaming/SslStreamTransport.cs:
```diff
@@ -0,0 +1,162 @@
+
+        public override string SchemeIdentifier
+        {
+            get { return "ssl"; }
```
Should be `ssl.tcp` or whatever the underlying transport is - see equivalent JVM code: https://github.com/akka/akka/blob/master/akka-remote/src/main/scala/akka/remote/transport/netty/NettyTransport.scala#L272

Aaronontheweb commented on src/core/Akka.Remote/Transport/Streaming/NetworkStreamTransport.cs:
```diff
@@ -0,0 +1,289 @@
+            Settings.ConfigureSocket(socket);
+
+            //TODO Does it even work if the port is not set?
+            int port = remoteAddress.Port ?? 2552;
```
No - blow it up if the remote port is not set.

Aaronontheweb commented on src/core/Akka.Remote/Transport/Streaming/StreamAssociationHandle.cs:
```diff
@@ -0,0 +1,238 @@
+        {
+            _writeQueue.Enqueue(payload);
+
+            // Always return true, we don't want to use the backoff in EndpointWriter
```
Using backoff might actually be the right thing to do in some cases, like if the writeQueue is full.

Aaronontheweb commented on src/core/Akka.Remote/Transport/Streaming/StreamAssociationHandle.cs:
```diff
@@ -0,0 +1,238 @@
+            _writeLoop = Task.Run(() => WriteLoop());
+        }
+
+        internal void Initialize(IHandleEventListener eventListener)
```
So correct me if I'm wrong - but does each instance of this transport use two separate threads perpetually? One for writing and another for reading respectively?

Or do the `WriteAsync` / `ReadAsync` calls use an I/O completion port and release those threads back until data is available / flushed? I'm assuming the latter but just want to be certain.

JeffCyr replied:
These are async methods, so you can view them as logical thread, they do not block a real thread. WriteAsync/ReadAsync use the IO threadpool.

The thread is released when await is called on an uncompleted task. When the task completes, the loop continue on the thread that completed the task. Might be an IO completion thread if coming back from ReadAsync/WriteAsync or might be a threadpool thread if coming back from the AsyncQueue.

Aaronontheweb replied:
Ok, cool - that's what I thought.

Aaronontheweb commented:
@JeffCyr general comment on the stream transports: this is something that I screwed up with Helios because I didn't know any better at the time, but we should pick a default Endianness in all of the buffers flushed onto the network and not let the underlying hardware decide. That's going to be especially true when .NET Core ships and Akka.NET can start running in a lot more places.

I recommend going with BigEndian just because that's the expected host network order for most protocols: http://stackoverflow.com/a/13514942/377476

Aaronontheweb commented:
Ok, finished with this review - some high level things:

1. Add some XML-DOCs to all of these classes so poor @sean-gilliam doesn't have to do it later :p 
2. Add some unit tests to verify that all of the settings classes correctly parse their HOCON values from a `Config` object.
3. I thought it was really clever how you used the `IAssociationEventListener` and `IHandleEventListener` to cut out a ton of boilerplate that it's in the Helios and DotNetty transports. I found your implementation much easier to follow than two of mine.

Another thing to consider is the threading part of this - how well does this transport implementation perform across X simultaneous connections? DotNetty and Helios try to push all of the dispatching work that happens from the transport onto a fixed size number of threads so we don't saturate the global `ThreadPool` as the number of connections grows. How is this transport going to deal with that and why might it be or not be an issue?

alexvaluyskiy commented:
@JeffCyr, have you compared performance with DotNetty transport from this PR #1465?

@Aaronontheweb, what the plans on Helios and DotNetty transports in the next release? Will DotNetty transport released?

JeffCyr commented:
@alexvaluyskiy I tried but there is currently a bug in DotNettyTransport or DotNetty itself that causes the throughput test to fail.

JeffCyr commented:
@Aaronontheweb I addressed your remarks in this [commit](https://github.com/akkadotnet/akka.net/pull/1826/commits/258c98dc97b1c3dc52a9f4bf4a2ae98a3bb03f15)

Here's my thought on some remarks:

##### `hostname` and `bind-hostname`
I followed the akka jvm settings here instead of helios. I also chose to bind to IPAny automatically unless localhost or a specific ip is specified in `hostname`. `bind-hostname` should rarely be used, `hostname` will work in most case, but it can be used to bind to a specific ip instead of IPAny.

##### Threading
The StreamTransports are designed to never block a thread. They will not cause a ThreadPool starvation, however some callbacks are called on the ThreadPool (e.g. when AsyncQueue.DequeueAsync is awaited). So the transport will be affected by a ThreadPool starvation if one occur in the current process. I will change the code to never hit the ThreadPool.

##### TODOs
- I will document the code thoroughly in another PR
- I will add 100% test coverage in a separate PR
